File: 90-subsets-ii/subsets-ii.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums); 
        List<List<Integer>> subsets = new ArrayList<>();
        subsets.add(new ArrayList<>());

        int start = 0, end = 0;

        for (int i = 0; i < nums.length; i++) {
            start = 0;

            if (i > 0 && nums[i] == nums[i - 1]) {
                start = end + 1;
            }

            end = subsets.size() - 1;
            int size = subsets.size();

            for (int j = start; j < size; j++) {
                List<Integer> subset = new ArrayList<>(subsets.get(j));
                subset.add(nums[i]);
                subsets.add(subset);
            }
        }

        return subsets;
    }
}

File: 90-subsets-ii/subsets-ii.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        def function(idx, path):
            if idx == len(nums):
                res.append(path[:])
                return
            
            path.append(nums[idx])
            function(idx+1, path)
            path.pop()

            while idx + 1 < len(nums) and nums[idx] == nums[idx + 1]:
                idx += 1

            function(idx + 1, path)

        function(0, [])

        return res   